boards/PDM/Src/Io/Io_Efuse.c

```diff
@@ -0,0 +1,821 @@
+    GPIO_TypeDef *ChipSelect_GPIO_Port,
+    uint16_t      ChipSelect_GPIO_Pin,
+    struct Efuse *e_fuse)
+{
+    ExitCode exit_code;
+    uint16_t command = 0x0000U;
+
+    // Place the register address into bits 10->13
+    command =
+        (uint16_t)((register_address & EFUSE_ADDR_MASK) << EFUSE_ADDR_SHIFT);
+    // Place register value to be written into bits 0->8
+    command = (uint16_t)(command | (register_value & EFUSE_SI_DATA_MASK));
+    // Invert watchdog bit state (Note: It is safe to do so even if the watchdog
+    // is disabled)
+    if (e_fuse->wdin_bit_to_set)
+    {
+        SET_BIT(command, WATCH_DOG_BIT);
+    }
+    else
+    {
+        CLEAR_BIT(command, WATCH_DOG_BIT);
+    }
+    // Invert watchdog bit state for next write
+    e_fuse->wdin_bit_to_set = !e_fuse->wdin_bit_to_set;
+
+    // Compute and set/clear parity value
+    Io_Efuse_SetParityBit(&command);
+
+    exit_code = Io_Efuse_WriteToEfuse(
+        &command, ChipSelect_GPIO_Port, ChipSelect_GPIO_Pin);
+
+    return exit_code;
+}
+
+static ExitCode Io_Efuse_ReadRegister(
+    uint8_t       register_address,
+    uint16_t *    register_value,
+    GPIO_TypeDef *ChipSelect_GPIO_Port,
+    uint16_t      ChipSelect_GPIO_Pin,
+    struct Efuse *e_fuse)
+{
+    ExitCode exit_code;
+    uint16_t command   = 0x0000U;
+    uint16_t tx_data[] = { 0x0000U };
+
+    // Place the Status Register address into bits 10->13
+    command =
+        (uint16_t)((SI_STATR_0_ADDR & EFUSE_ADDR_MASK) << EFUSE_ADDR_SHIFT);
+    // Shift bit 3 of the address (SOA3: the channel number) to the 13th bit
+    command = (uint16_t)(
+        command | ((register_address & SOA3_MASK) << EFUSE_ADDR_SHIFT) |
+        (register_address & (SOA2_MASK | SOA1_MASK | SOA0_MASK)));
+    // Invert watchdog bit state (Note: It is safe to do so even if the watchdog
+    // is disabled)
+    if (e_fuse->wdin_bit_to_set)
+    {
+        command |= WATCH_DOG_BIT;
+    }
+    else
+    {
+        command &= ~(WATCH_DOG_BIT);
+    }
+    // Invert watchdog bit state for next write
+    e_fuse->wdin_bit_to_set = !e_fuse->wdin_bit_to_set;
+
+    // Compute and set/clear parity bit
+    Io_Efuse_SetParityBit(&command);
+
+    tx_data[0] = command;
+    exit_code  = Io_Efuse_ReadFromEfuse(
+        tx_data, register_value, ChipSelect_GPIO_Port, ChipSelect_GPIO_Pin);
+    // Only return register contents and clear bits 9->15
+    CLEAR_BIT(*register_value, ~EFUSE_SO_DATA_MASK);
+
+    return exit_code;
+}
+
+static void Io_Efuse_SetParityBit(uint16_t *spi_command)
+{
+    uint16_t spi_data = *spi_command;
+    uint8_t  parity_bit;
+    // Calculate parity for bits in spi command
+    for (parity_bit = 0; spi_data > 0; spi_data >>= 1)
+    {
+        parity_bit ^= (spi_data & 1);
+    }
+
+    if (parity_bit == 1)
+    {
+        SET_BIT(*spi_command, PARITY_BIT);
+    }
+    else
+    {
+        CLEAR_BIT(*spi_command, PARITY_BIT);
+    }
+}
+
+static ExitCode Io_Efuse_WriteToEfuse(
```

Comment:
Weren't these supposed to be implemented in a shared SPI driver?

Reply:
Hmmm so currently the only other board that has SPI code is the DIM and it's data msg is slightly different. The DIM sends out 3 x 8-bit messages while PDM sends out 1 16-bit msg hence the typecasting to 8-bits here and the 1U parameter. Lets revisit a shared library once we add the SPI driver code for the BMS? I also wanted to try SPI in one of the non-blocking methods (interrupt or DMA) once the new PDM is made.

Reply:
If we were to have a shared wrapper function for SPI it might look something like this:
```
ExitCode Shared_WriteSPIData(
    uint16_t *    TxData,
    uint8_t         messages,
    GPIO_TypeDef *GPIO_Port,
    uint16_t      GPIO_Pin)
{
    HAL_StatusTypeDef status;

    HAL_GPIO_WritePin(GPIO_Port, GPIO_Pin, GPIO_PIN_RESET);
    status = HAL_SPI_TransmitReceive(
        efuse_spi_handle, (uint8_t *)TxData, (uint8_t *)TxData, messages, 100U);
    HAL_GPIO_WritePin(GPIO_Port, GPIO_Pin, GPIO_PIN_SET);

    if (status != HAL_OK)
    {
        return EXIT_CODE_TIMEOUT;
    }

    return EXIT_CODE_OK;
}

ExitCode Shared_ReadSPIData(
    uint16_t *    tx_data,
    uint16_t *    rx_data,
    uint8_t         messages,
    GPIO_TypeDef *chip_select_port,
    uint16_t      chip_select_pin)
{
    HAL_StatusTypeDef status;

    // Send command to read from status register
    // Data is returned on the following SPI transfer
    HAL_GPIO_WritePin(chip_select_port, chip_select_pin, GPIO_PIN_RESET);
    status = HAL_SPI_TransmitReceive(
        efuse_spi_handle, (uint8_t *)tx_data, (uint8_t *)rx_data, messages, 100U);
    HAL_GPIO_WritePin(chip_select_port, chip_select_pin, GPIO_PIN_SET);

    if (status != HAL_OK)
    {
        return EXIT_CODE_TIMEOUT;
    }

    // Receive data from E-fuse by sending dummy data
    *tx_data = 0xFFFF;

    HAL_GPIO_WritePin(chip_select_port, chip_select_pin, GPIO_PIN_RESET);
    status = HAL_SPI_TransmitReceive(
        efuse_spi_handle, (uint8_t *)tx_data, (uint8_t *)rx_data, messages, 100U);
    HAL_GPIO_WritePin(chip_select_port, chip_select_pin, GPIO_PIN_SET);

    if (status != HAL_OK)
    {
        return EXIT_CODE_TIMEOUT;
    }

    return EXIT_CODE_OK;
}
```